### ISSUE TITLE
Auto-Visibility of horizontal scrollbar

### ISSUE THREAD

**Issue by RaeuberHotzenplotz** [jacobslusser/ScintillaNET#216]
I'm looking for an option that makes the horizontal scrollbar only visible if the width of the currently visible text (or the entire text) exceeds the visible area. It would be possible to implement this behavior manually by calculating the text width when the control is resized or the text is changed by setting the HScrollBar-Property. That leaves the problem of measuring the text bounds in pixels, for which i didn't find any methods in the current version. Who can help?

**Comment by jacobslusser** [comment]
Scintilla has a long history with this feature/issue. The best you can do set

```cs
scintilla.ScrollWidth = 0;
scintilla.ScrollWidthTracking = true; // the default
```

Setting `ScrollWidth` to `0` will reset the way Scintilla tracks the width of text in the document. With `ScrollWidthTracking` set to `true`, Scintilla will automatically increase the horizontal scroll range as it encounters longer lines.

For the most part this is what you want. There is one caveat though.... Scintilla will not automatically *reduce* the horizontal scroll range when it no longer needs to be so wide.

The reasons for this are many but it basically comes down to performance. The creator of Scintilla (Neil) has long held the position that to implement such a feature would be detrimental to performance. Thus Scintilla has a lazy approach to calculating the horizontal scroll width. As it renders lines of text that are too long it increases the scroll width; but to do the reverse and shrink the width when the longest line gets shorter would require remeasuring each line to determine the new longest line... It's complicated, but that's why.

My recommendation is to simply set the `ScrollWidth` to `0` each time a new document is loaded into the control. If you want to get fancy you can try to reset that value sometime during the life of the document being edited, but I personally wouldn't bother.

**Comment by robmburke** [comment]
And yet the IDEs I use don't seem to have this problem.

**Comment by jacobslusser** [comment]
They also have a staff of paid programmers. ;)

**Comment by RaeuberHotzenplotz** [comment]
For some reason, setting ScrollWidth to 0 (or any other number) immediately sets it back to 2000.

**Comment by jacobslusser** [comment]
Sorry, I gave you bad advice... don't use `0`, Scintilla interprets that as null and thus does nothing to change the `ScrollWidth`. Try setting it to `1`. I've tested this and it works, but if you also have `ScrollWidthTracking` enabled Scintilla will increase any value you specify to the minimum necessary for the text displayed in the control. In my testing I had a little "Hello World" in the control and so setting `ScrollWidth` to `1` would remove the scrollbar (which wasn't necessary to display "Hello World") but getting the`ScrollWidth` property immediately after I saved it would return a value of `174` -- because that was the minimum necessary for "Hello World" and I had `ScrollWidthTracking` on. If you want complete control of the scroll bar, turn `ScrollWidthTracking` off. 

Just to repeat, a value of `0` will not work because it is a no-op to Scintilla. But any other small value should work.
